Readme.md
Expo
====

Expo is a meta-framework for rapid [Node.js] web development. It provides you 
will the sensible defaults you need to build the backend of a modern web 
application.

``` cmd
# Install it via:
$ npm install -g expo
$ expo --help
```

 * __Routing__
   * Routing API *-- via [Express]*
   * Middleware support *-- via [connect]*
   * Sessions, helpers, cookies
 * __Assets__
   * Asset pipeline *-- via [connect-assets]*
   * Concatenation, compression, cache-busting
   * Support for CoffeeScript, Less, Stylus
 * __Database__
   * Database ORM *-- via [Sequelize]*
   * Database migrations
 * __Testing__
   * Tests *-- via [Mocha]*
   * Assertions *-- via [Chai]*
   * Integration tests *-- via [superagent]*
 * __Utilities__
   * Command-line tasks *-- via [commander.js]*
   * Auto-restarting on changes *-- via [supervisor]*
   * Logging
   * Config file loading

[![Status](https://travis-ci.org/expo-js/expo.png?branch=master)](https://travis-ci.org/expo-js/expo)  [![NPM version](https://badge.fury.io/js/expo.png)](http://badge.fury.io/js/expo)

## Your first app

Expo lets you start up sites very fast.
Create a project using the `expo` command. Here we'll use `--bare`, which
generates the bare minimum:

```
$ expo --bare hello

        hello/  your project
                Readme.md
                run
                app.js
                .gitignore
                package.json

 initializers/  code that initializes the app environment
                app/initializers/app.js

       routes/  auto-loaded files that create URL routes
                app/routes/home.js

        views/  template files
                app/views/index.jade
                app/views/layout.jade

  install dependencies:
    $ cd hello
    $ npm install && npm shrinkwrap

  run the app:
    $ ./run server

  more info:
    $ ./run --help
```

[Read about creating projects >]( guide.html#expo-executable )

## Full setup

...But this isn't always that useful, since you probably will need assets,
databases, sessions, tests, helpers and other doodads. Use `expo` without 
switches to generate the default layout:

```
$ expo hello

        hello/  your project
                Readme.md
                run
                app.js
                .gitignore
                package.json

       assets/  where your images, CSS and JS files live
                app/assets/css/application.styl
                app/assets/js/application.js
                app/assets/img/.gitkeep

 initializers/  code that initializes the app environment
                app/initializers/app.js
                app/initializers/assets.js
                app/initializers/sessions.js

      helpers/  functions available in your views
                app/helpers/app_helper.js

       routes/  auto-loaded files that create URL routes
                app/routes/home.js

        tasks/  command line tasks
                app/tasks/db.js

        views/  template files
                app/views/index.jade
                app/views/layout.jade

       config/  yaml config files
                config/database.yml.example
                config/secret_token.yml

           db/  development sqlite database
                db/.gitkeep

         lib/   models and code
                lib/.gitkeep

       public/  normal files go here
                public/robots.txt

         test/  tests, specs
                test/setup.js
                test/app_test.js

  install dependencies:
    $ cd aoeu
    $ npm install && npm shrinkwrap
```

[Read about creating projects >]( guide.html#expo-executable )

## Command line runner

Expo projects come with a runner that gives you access to default tasks.
Simply invoke ./run in a project.

```
$ cd myproject
$ ./run --help

  Usage: run [options] [command]

  Commands:

    server [port] [..]     Starts the server (alias: "s")
    console                Opens a console (alias: "c")
    runner [cmd]           Runs a command (alias: "r")
    db-create              Creates the environment's database
    db-drop                Drops the environment's database
    db-migrate             Run database migrations
    gen-migration [name]   Creates a migration file in migrations/
    assets-precompile      Precompiles asset files

  Options:

    -h, --help       output usage information
    -V, --version    output the version number
    -e, --env [env]  Environment to start in [develompent]
```

[Read about the runner >]( guide.html#runner )

## Server

Run your a project using `./run server` (or its shortcut: `./run s`).
In development mode, you server will automatically reload on code changes.

```
$ ./run s

  => Auto-restarting on changes
  => Running development mode at http://0.0.0.0:4567
  => Ready [290ms]
  GET / 200 486ms - 241b
  GET / 200 116ms - 241b
  GET /css/application.css 200 10ms - 36b
  GET /js/application.js 200 2ms - 1b
```

## Console

And you get a convenient REPL console.

```js
$ ./run c

  > 2 + 4
  6
  > app.get('env')
  'development'
```


## Entry point

Your Expo app behaves just like any Node package would. It has a main entry
point, `app`, which is a Connect app.

```js
$ node

  > app = require('.')
  { ... }
  > app.get('env')
  'development'
  > app.conf('database').username
  'youruser'
  > http.createServer(app).listen(5678);
```


## Tests (via mocha)

Tests are available via Mocha and Chai. Just like all Node packages, the command 
is the standard `npm test`.

```
$ npm test

  > myapp@0.1.0 test
  > ./node_modules/.bin/mocha -R spec

  >  App loaded for test environment

    App
      Homepage should work (799ms)
      should have the right env

    2 tests complete (807 ms)
```

[Read about tests >]( guide.html#tests )


How it looks
============

## Initializers

Here's the main entry point, and the main initializer.
You can dump JS files in `app/initializers/` and they will be automatically
loaded when the app is loaded.

```js
$ cat app.js

  var app = module.exports = require('express')();
  require('expo')(app, __dirname);


$ cat app/initializers/app.js

  var express = require('express');

  module.exports = function(app) {
    app.set('view engine', 'jade');
    app.use(express['static'](app.path('public')));
    app.use(express.methodOverride());
    app.use(express.bodyParser());

    app.configure('development', function() {
      app.use(express.favicon());
      app.use(express.logger('dev'));
      app.use(express.errorHandler());
    });
  };
```


## Routes

Routes in app/routes are automatically loaded.

```js
$ cat app/routes/home.js

  module.exports = function(app) {
    app.get('/', function(req, res) {
      res.render('index', {});
    });
  };
```

[Read about routes >]( guide.html#routes )


## Helpers

These will be available in your views automatically.

```js
$ cat app/helpers/name_helper.js

  module.exports = {
    hello: function(name) {
      return "Hello " + name;
    }
  };

$ cat app/views/hello.jade

  div= hello("Rico")
```

[Read about helpers >]( guide.html#helpers )


## Configuration

Expo provides you helpers for Yaml and JSON configuration files via
`app.conf()`.

```js
$ cat config/database.yml

  production:
    username: 'rsc'
    dbname: 'foobar_production'
  development:
    username: 'rsc'
    dbname: 'foobar_development'

$ ./run console

  > app.conf('database').dbname
  "foobar_development"
```


Misc
====


Acknowledgements
----------------

© 2013, Rico Sta. Cruz. Released under the [MIT License].

[Image] was taken from Flickr, licensed under Creative Commons.

**Expo** is authored and maintained by [Rico Sta. Cruz] with help
from its [contributors].

 * [My website](http://ricostacruz.com) (ricostacruz.com)
 * [Github](http://github.com/rstacruz) (@rstacruz)
 * [Twitter](http://twitter.com/rstacruz) (@rstacruz)

[MIT License]: http://www.opensource.org/licenses/mit-license.php
[Rico Sta. Cruz]: http://ricostacruz.com
[contributors]: http://github.com/rstacruz/expo/contributors

[connect-assets]: https://github.com/adunkman/connect-assets
[Express]: http://expressjs.com/
[connect]: https://github.com/senchalabs/connect
[Node.js]: http://nodejs.org/
[commander.js]: https://github.com/visionmedia/commander.js
[Sequelize]: http://www.sequelizejs.com
[Mocha]: http://visionmedia.github.io/mocha
[supervisor]: https://github.com/isaacs/node-supervisor
[Chai]: http://chaijs.com/
[Chai should]: http://chaijs.com/guide/styles/#should
[Chai assert]: http://chaijs.com/api/assert/
[Superagent]: https://npmjs.org/package/superagent
[Sinon.js]: http://sinonjs.org/
[Image]:http://www.flickr.com/photos/feuilllu/7473428030/sizes/l/in/photostream/


History.md
## v0.1.7 - August 30, 2013

 * Template: add 'log errors' to production.
 * Log errors.

## v0.1.6 - August 28, 2013

 * Add `peerDependencies` to the NPM packages.
 * Template: create a `Procfile` for Heroku.

## v0.1.5 - June 6, 2013

 * New `throw errors` setting -- when set to true, 500 error pages will not 
 show.

 * Source map support via the `source-map-support` package, enabled by default.

## v0.1.4 - June 17, 2013

Expo:

   * Console: it now triggers the 'console' event allowing you to hook things, 
   allowing you to hook

   * App: implement `app.loadConsole()`. This allows you to run `app = 
   require('app').loadConsole()` in any REPL shell (like `coffee`).

   * There's now a default `uncaughtException` handler to terminate the 
   application when one happens.

Template:

   * The `package.json` now gives you the command `npm run autotest` to 
   continuously run tests.

## v0.1.3 - June 13, 2013

Expo:

 * Config: allow 'default' configuration scope.
 * CLI: improve `expo --help` help text.
 * Add support for 404 and 500 error handlers.
 * Log timestamps in production mode.
 * Template: consolidate sessions into the main initializer.
 * App: add `add.emit()` function.

## v0.1.2 - May 30, 2013

App generator:

 * New projects now have a `repository` in package.json.
 * New projects now default to version "0.0.1".
 * Fix the default tasks.
 * Fix the 'npm install' error.
 * Automatically gitignore compiled assets.
 * Move test packages into devDependencies.
 * Remove pg/mysql by default.

Expo:

 * CLI tasks that aren't returning JS code are now ignored.

## v0.1.1 - May 30, 2013

Initial feature-complete release.


index.js
var expo = module.exports = function(app, dir) {
  // Use coffeescript right away to allow .coffee files
  require('coffee-script');
  // Use source maps
  require('source-map-support').install();

  require('./lib/app')(app);
  app.root = dir;
  return app;
};

/**
 * Middleware proxy to express.errorHandler, but let 404 errors through.
 *
 *     app.use(app.errorHandler(express.errorHandler()));
 */

expo.errorHandler = function(handler) {
  return function(err, req, res, next) {
    if (err === 404) return next(404);
    handler(err, req, res, next);
  };
};

/**
 * Middleware for logging errors to STDERR.
 * This is activated when using `app.set('log errors', true);`.
 */

expo.errorLogger = function(err, req, res, next) {
  if (err === 404) return next(err);

  var addr = req.socket && (req.socket.remoteAddress || (req.socket.socket && req.socket.socket.remoteAddress));
  var ver  = req.httpVersionMajor + '.' + req.httpVersionMinor;

  console.error("");
  console.error("ERROR:");
  console.error("    ", req.method, req.url, "HTTP/"+ver, "-", res.statusCode);
  console.error("    ", "Remote-IP:", addr);
  console.error("    ", "Date:", (new Date()).toUTCString());

  // Make connect show the errors
  next(err);
};


lib/app.js
var path = require('path');
var loadModules = require('./path_helpers').loadModules;
var EventEmitter = require('events').EventEmitter;

/**
 * Application extensions mixin.
 * Extends an Express app with more functions.
 */

var app = module.exports = function(app) {

  var expo = require('..');

  /**
   * Event emitter. Used by `emit()` and such.
   * @api private
   */

  var events = new EventEmitter();

  /**
   * The root path. To be overriden later.
   */

  app.root = '';

  /**
   * Gets a path relative to the root.
   *
   *     app.path('public')
   *     app.path('app/migrations')
   */

  app.path = function() {
    return path.resolve.apply(this, [this.root].concat([].slice.call(arguments)));
  };

  var loaded = false;

  /**
   * Loads all files needed to run an Express server.
   *
   *     app.load();
   *     app.listen(3000);
   *
   * Also emits the following events in this sequence:
   *
   *  - load:before
   *  - initializers:before, initializers:after
   *  - helpers:before, helpers:after
   *  - routes:before, routes:after
   *  - load:after
   */

  app.load = function(env) {
    if (!loaded) {
      process.chdir(app.root);

      // Set environment if asked (usually test).
      if (env) {
        process.env.NODE_ENV = env;
        app.set('env', env);
      }

      env = app.get('env');
      if (env !== 'development') app.log = require('./logger')(env);

      // Hooks: do pre-load hooks that extensions may listen for.
      app.emit('load:before');
      if (env === 'test') app.emit('load:test:before');

      // Load initializers of the application.
      loadAppPath('initializers', 'function', function(mixin) { mixin(app); });

      // Middleware for 'file not found' and 'server error' -- must always be
      // the last
      app.use(notFound);
      app.use(expo.errorLogger);
      app.use(serverError);

      // Apply the helpers using `.local()` to make them available to all views.
      loadAppPath('helpers', 'object', function(helpers) { app.locals(helpers); });

      // Load routes of the application.
      loadAppPath('routes', 'function', function(mixin) { mixin(app); });

      // After hooks
      if (env === 'test') app.emit('load:test:after');
      app.emit('load:after');

      loaded = true;
    }

    return this;
  };

  /**
   * Loads the app in a console-friendly way. Environment optional.
   */

  app.loadConsole = function(env) {
    app.load(env);
    app.emit('console');
    global.app = app;
  };

  var command;

  /**
   * Returns the commander command line parser.
   *
   *      app.cli()
   *      app.cli().parse(...)
   */

  app.cli = function() {
    if (!command) {
      command = require('commander');

      // Import [1] default tasks, [2] extensions tasks, [3] app tasks.
      require('./cli')(app, command);
      app.emit('cli', command);
      loadModules(app.path('app/tasks'), 'function', function(mixin) { mixin(app, command); });
    }

    return command;
  };

  // Logger
  // ------

  /**
   * Simple logging facility.
   *
   *     app.log.debug('Loading models');
   *     app.log.info('Loading models');
   */

  app.log = require('./logger')(process.env.NODE_ENV || 'development');

  // Events
  // ------

  /**
   * The events.
   * You may emit events here.
   */

  app.events = events;

  /**
   * Listens to a given event.
   *
   *     app.on('load:before', function() { ... })
   *
   * This emits events during the load process, allowing extensions to hook
   * onto certain parts of the load process.
   *
   * This is just shorthand for `app.events.on(x, ...)`.
   *
   * See `app.load()` for a catalog of events that it emits.
   */

  app.on = function(eventName, listener) {
    events.on(eventName, listener);
    return this;
  };

  /**
   * Emits an event.
   *
   *     app.emit('load:before');
   *     app.emit('cli', app.cli);
   */

  app.emit = function(eventName, arg) {
    events.emit(eventName, app, arg);
    return this;
  };

  // Config
  // ------
  
  var configData = {};

  /**
   * Loads configuration from a file. Yaml, JSON and JS are supported.
   *
   *     // Reads `config/secret_token.yml`
   *     // (or .json, or .js, or .coffee)
   *     app.conf('secret_token');
   */

  app.conf = function(file) {
    if (!(file in configData)) {
      var env = app.get('env');
      var load = require('./conf');

      var data = load(file, [
        app.path('config'),
        path.join(process.cwd(), 'config')
      ]);

      configData[file] = (env in data ? data[env] : data['default']);
    }

    return configData[file];
  };

  // Defaults
  // --------

  /**
   * Set some defaults
   */

  app.set('views', app.path('app/views'));

  // Private helpers
  // ---------------

  /**
   * Loads mixins in a given path.
   * @api private
   */

  function loadAppPath(path, type, callback) {
    events.emit(path+':before', app);
    loadModules(app.path('app', path), type, callback);
    events.emit(path+':after', app);
  }

  function notFound(req, res, next) {
    next(404);
  }

  function serverError(err, req, res, next) {
    if (err === 404) {
      res.status = 404;
      res.statusCode = 404;
      res.render('errors/404', { status: 404, url: req.url });
      return;
    }

    if (app.get('show errors')) {
      res.render('errors/500', { status: 500, error: err });
    }

    next(err);
  }
};
